test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,146 @@
+#include "test/test_common/network_utility.h"
+#include "test/test_common/simulated_time_system.h"
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+using testing::AssertionFailure;
+using testing::AssertionResult;
+using testing::AssertionSuccess;
+using testing::IsSubstring;
+
+namespace Envoy {
+namespace {
+
+const std::string kConfig = R"EOF(
+admin:
+  access_log_path: /dev/null
+  address:
+    socket_address:
+      address: 127.0.0.1
+      port_value: 0
+)EOF";
+
+class CdsIntegrationTest : public XdsIntegrationTestBase,
+                           public Grpc::GrpcClientIntegrationParamTest {
+public:
+  CdsIntegrationTest()
+      : XdsIntegrationTestBase(Http::CodecClient::Type::HTTP2, ipVersion(), kConfig) {}
+
+  void TearDown() override {
+    AssertionResult result = xds_connection_->close();
+    RELEASE_ASSERT(result, result.message());
+    result = xds_connection_->waitForDisconnect();
+    RELEASE_ASSERT(result, result.message());
+    xds_connection_.reset();
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  envoy::api::v2::Cluster buildCluster(const std::string& name) {
+    return TestUtility::parseYaml<envoy::api::v2::Cluster>(
+        fmt::format(R"EOF(
+      name: {}
+      connect_timeout: 5s
+      type: STATIC
+      load_assignment:
+        cluster_name: {}
+        endpoints:
+        - lb_endpoints:
+          - endpoint:
+              address:
+                socket_address:
+                  address: {}
+                  port_value: {}
+      lb_policy: ROUND_ROBIN
+      http2_protocol_options: {{}}
+    )EOF",
+                    name, name, Network::Test::getLoopbackAddressString(ipVersion()),
+                    fake_upstreams_[0]->localAddress()->ip()->port()));
+  }
+
+  void initializeCds() {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      // Setup cds and corresponding gRPC cluster.
+      auto* cds_config =
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->mutable_api_config_source();
+
+      cds_config->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+      cds_config->mutable_request_timeout()->set_seconds(1);
+      auto* grpc_service = cds_config->add_grpc_services();
+      setGrpcService(*grpc_service, "my_cds_cluster", fake_upstreams_[0]->localAddress());
+      grpc_service->mutable_envoy_grpc()->set_cluster_name("my_cds_cluster");
+
+      auto* cds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      cds_cluster->set_name("my_cds_cluster");
+      cds_cluster->mutable_connect_timeout()->set_seconds(5);
+      auto* sockaddr = cds_cluster->add_hosts()->mutable_socket_address();
+      sockaddr->set_protocol(envoy::api::v2::core::SocketAddress::TCP);
+      sockaddr->set_address(Network::Test::getLoopbackAddressString(ipVersion()));
+      sockaddr->set_port_value(0);
+      cds_cluster->clear_http2_protocol_options();
```

Comment:
Nit: I don't think this clear is needed.

Reply:
Done.